src/input/keyboard.c

```diff
@@ -447,58 +447,36 @@ handle_menu_keys(struct server *server, struct keysyms *syms)
 	}
 }
 
-static void
-toggle_direction(enum lab_cycle_dir *direction)
-{
-	if (*direction == LAB_CYCLE_DIR_FORWARD) {
-		*direction = LAB_CYCLE_DIR_BACKWARD;
-	} else if (*direction == LAB_CYCLE_DIR_BACKWARD) {
-		*direction = LAB_CYCLE_DIR_FORWARD;
-	}
-}
-
-static void
+/* Returns true if the keystroke is consumed */
+static bool
 handle_cycle_view_key(struct server *server, struct keyinfo *keyinfo)
 {
 	for (int i = 0; i < keyinfo->translated.nr_syms; i++) {
 		if (keyinfo->translated.syms[i] == XKB_KEY_Escape) {
 			/* cancel view-cycle */
 			osd_finish(server);
-			return;
+			return true;
 		}
 	}
 
+	if (keyinfo->is_modifier) {
+		return false;
+	}
+
 	/* cycle to next */
-	if (!keyinfo->is_modifier) {
-		enum lab_cycle_dir direction = server->osd_state.initial_direction;
-		for (int i = 0; i < keyinfo->translated.nr_syms; i++) {
-			if (keyinfo->translated.syms[i] == XKB_KEY_Up
-					|| keyinfo->translated.syms[i] == XKB_KEY_Left) {
-				direction = LAB_CYCLE_DIR_BACKWARD;
-				goto miss_shift_toggle;
-			}
-			if (keyinfo->translated.syms[i] == XKB_KEY_Down
-					|| keyinfo->translated.syms[i] == XKB_KEY_Right) {
-				direction = LAB_CYCLE_DIR_FORWARD;
-				goto miss_shift_toggle;
-			}
+	for (int i = 0; i < keyinfo->translated.nr_syms; i++) {
+		if (keyinfo->translated.syms[i] == XKB_KEY_Up
```

Comment:
Looks like we could combine this loop with the `Escape` one on top and move the `is->modifier` check before that one.

Reply:
Nice suggestion. Fixed.